dll/win32/shell32/dialogs/filetypes.cpp

```diff
@@ -1462,129 +1636,166 @@ EditTypeDlgProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
 static INT CALLBACK
 FileTypesDlg_CompareItems(LPARAM lParam1, LPARAM lParam2, LPARAM lParamSort)
 {
-    PFILE_TYPE_ENTRY Entry1, Entry2;
-    int x;
-
-    Entry1 = (PFILE_TYPE_ENTRY)lParam1;
-    Entry2 = (PFILE_TYPE_ENTRY)lParam2;
-
-    x = wcsicmp(Entry1->FileExtension, Entry2->FileExtension);
-    if (x != 0)
-        return x;
+    PFILE_TYPE_GLOBALS pG = (PFILE_TYPE_GLOBALS)lParamSort;
+    PFILE_TYPE_ENTRY p1 = (PFILE_TYPE_ENTRY)lParam1, p2 = (PFILE_TYPE_ENTRY)lParam2;
+    int x = 0;
+    if (pG->SortCol == 1)
+        x = wcsicmp(GetTypeName(p1, pG), GetTypeName(p2, pG));
+    if (!x && !(x = p1->IsExtension() - p2->IsExtension()))
+        x = wcsicmp(p1->FileExtension, p2->FileExtension);
+    return x * pG->SortReverse;
+}
 
-    return wcsicmp(Entry1->FileDescription, Entry2->FileDescription);
+static void
+FileTypesDlg_Sort(PFILE_TYPE_GLOBALS pG, HWND hListView, INT Column = -1)
+{
+    pG->SortReverse = pG->SortCol == Column ? pG->SortReverse * -1 : 1;
+    pG->SortCol = Column < 0 ? 0 : (INT8) Column;
+    ListView_SortItems(hListView, FileTypesDlg_CompareItems, (LPARAM)pG);
 }
 
 static VOID
 FileTypesDlg_InitListView(HWND hwndDlg, HWND hListView)
 {
     RECT clientRect;
+    LPCWSTR columnName;
+    WCHAR szBuf[50];
+
     LVCOLUMNW col;
-    WCHAR szName[50];
-    DWORD dwStyle;
-    INT columnSize;
+    col.mask = LVCF_WIDTH | LVCF_TEXT | LVCF_SUBITEM | LVCF_FMT;
+    col.fmt = 0;
+
+    GetClientRect(hListView, &clientRect);
+    INT column0Size = (clientRect.right - clientRect.left) / 4;
 
-    if (!LoadStringW(shell32_hInstance, IDS_COLUMN_EXTENSION, szName, _countof(szName)))
+    columnName = L"Extensions"; // Default to English
+    if (LoadStringW(shell32_hInstance, IDS_COLUMN_EXTENSION, szBuf, _countof(szBuf)))
     {
-        // default to english
-        wcscpy(szName, L"Extensions");
+        columnName = szBuf;
+        szBuf[_countof(szBuf) - 1] = UNICODE_NULL; // Make sure it's null terminated
```

Comment:
The LoadString function _**will**_ already null-terminate it, the way you use the function:
https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-loadstringw
> Loads a string resource from the executable file associated with a specified module and _either copies the string into a buffer **with a terminating null character**_ or returns a read-only pointer to the string resource itself.
> [...]
> [in] cchBufferMax
Type: int
The size of the buffer, in characters. _The string is truncated **and null-terminated** if it is longer than the number of characters specified._ [...]